### PR TITLE
libiconv: workarounds to support conan v2

### PR DESCRIPTION
**Pull request by jcar87**
Specify library name and version:  **libiconv/all**

Workarounds to support cross-compatibility of this recipe between Conan 1.x and Conan 2.0:

* Avoid comparing `self.settings.compiler` directly, but rather map the version to compare depending on the current compiler
* Avoid calling the legacy `get_gnu_triplet`, in practice the triplet differs only in the architecture

### CODE REVIEW

**Review by czoido — APPROVED** [review]

**Comment by jcar87** [comment]
> > Avoid calling the legacy get_gnu_triplet, in practice the triplet differs only in the architecture
> 
> I understand the workaround to fully support conan v2, but wouldn't it be better to fix AutotoolsToolchain instead in conan client? Default triplets are wrong for msvc ([conan-io/conan#7460](https://github.com/conan-io/conan/issues/7460)).

Indeed, we are currently evaluating the next steps surrounding this and other related issues. Even once that happens, it will still take a while for the user base and Conan Center CI service to be updated - this just changes the way the current workaround is implemented so as to not impede Conan 2 usage, hopefully without impacting those users that are cross-building building on Windows (which I also understand is not the most common use case).

File: recipes/libiconv/all/conanfile.py [hunk 1]
```diff
@@ -85,17 +84,23 @@ def generate(self):
...
 
         tc = AutotoolsToolchain(self)
-        if (self.settings.compiler == "Visual Studio" and Version(self.settings.compiler.version) >= "12") or \
-           (self.settings.compiler == "msvc" and Version(self.settings.compiler.version) >= "180"):
+        msvc_version = {"Visual Studio": "12", "msvc": "180"}
+        if is_msvc(self) and Version(self.settings.compiler.version) >= msvc_version[str(self.settings.compiler)]:
+            # https://github.com/conan-io/conan/issues/6514
             tc.extra_cflags.append("-FS")
         if cross_building(self) and is_msvc(self):
+            triplet_arch_windows = {"x86_64": "x86_64", "x86": "i686", "armv8": "aarch64"}
```
Comment (SpaceIm) [inline recipes/libiconv/all/conanfile.py:92]: cross-build to arm 32 bits is not handled anymore

**Comment by prince-chrismc** [comment]
What about making the existing Conan 2.0 implementation thats being used public https://github.com/conan-io/conan/blob/904162523b2947998919bb6a84ece6ec26e5730e/conan/tools/gnu/get_gnu_triplet.py#L4?

File: recipes/libiconv/all/conanfile.py [hunk 2]
```diff
@@ -15,7 +15,6 @@
 from conan.tools.layout import basic_layout
 from conan.tools.microsoft import is_msvc, unix_path
 from conan.tools.scm import Version
-from conans.tools import get_gnu_triplet
```
Comment (SSE4) [inline recipes/libiconv/all/conanfile.py:18]: why not use `get_gnu_triplet`? is it buggy or something?
repeating the same logic in every recipe doesn't scale

**Review by prince-chrismc — APPROVED** [review]
2.0 priority... but let's see what solution the client gets before repeating this everywhere

File: recipes/libiconv/all/conanfile.py (hunk 2 above)
Comment (prince-chrismc) [reply recipes/libiconv/all/conanfile.py:18]: It's not migrated to 2.0 yet 🙃

Comment (SSE4) [reply recipes/libiconv/all/conanfile.py:18]: let's create an issue to migrate it then.
every old v1 `tools.*` should have a drop-in replacement in v2, or at very least, some clear migration path how to replace it.
let's avoid adding workarounds that don't scale to recipes.

**Review by SSE4 — CHANGES_REQUESTED** [review]
adding triplet mapping to every recipe doesn't scale.
this logic should be internal to `get_gnu_triplet`.

**Comment by SSE4** [comment]
why does it merge with changes requested :thinking: ?

**Comment by prince-chrismc** [comment]
Community reviewers can not block PR 🙀

**Comment by SSE4** [comment]
> Community reviewers can not block PR scream_cat

so, any point of having community reviewed if they are simply ignored?

**Comment by jcar87** [comment]
> 2.0 priority... but let's see what solution the client gets before repeating this everywhere

We are currently working on a fix for https://github.com/conan-io/conan/issues/7460, which is the underlying issue in this instance, and will address the issue in the internal implementation as @SSE4  rightly suggests. Once it is addressed, this logic can go away altogether. It should also be more robust for the 70-80 recipes that fall under the same umbrella (windows+msvc+autotools), for those cases in which users are cross-compiling.

As for scalability, I absolute agree that this solution does not scale to a big number of recipes. However, at the moment there are only two recipes that use this specific pattern, the other one being `icu`, so I don't expect this to cause a cascade of workarounds to propagate elsewhere, and has very limited impact. 

Apologies @SpaceIm , you're absolutely right about the lack of handling of 32-bit ARM, I just didn't get to address it in time before it got merged by the bot. Given that we are planning to address the issue in Autotools, I expect to have a long term solution in place before this causes practical disruption (especially considering this recipe does not successfully cross-build due to issues you've already identified with the resource compiler building objects for the wrong architecture).

**Pull request merged**